---
title: "scRNABatchQC Report"
output: 
  rmdformats::material:
    self_contained: false
    thumbnails: true
    lightbox: true
    gallery: true
    highlight: tango  
params:
  data: plotData
  displayFigure: TRUE
---

```{r setup, include=FALSE}
library(RColorBrewer)
library(knitr)
library(reshape2)
library(ggplot2)
library(data.table)
library(DT)
library(RCurl)
library(htmltools)

knitr::opts_chunk$set(echo = TRUE, autodep=TRUE, cache.comments=FALSE, message=FALSE, warning=FALSE)

addLinkTag<-function(text,link) {
  result<-paste0("<a href='",link,"' target='_blank'>",text,"</a>")
  return(result)
}

figRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("figcap.prefix"), 
           sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight")) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
             " ", caption)
    } else {
      used[label] <<- TRUE
      paste(prefix, tag[label])
    }
  }
})
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")

tabRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("tabcap.prefix"), 
           sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight")) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
             " ", caption)
    } else {
      used[label] <<- TRUE
      paste(prefix, tag[label])
    }
  }
})
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")
```

```{r data, include=FALSE, cache=FALSE}
DISPLAY_FIGURE=params$displayFigure
plotData<-params$data
sces <- plotData$sces
scesall <- plotData$scesall
diffFC <- plotData$diffFC
hvgPathways<-plotData$hvgPathways
pc1Pathways<-plotData$pc1Pathways

if(length(sces) < 9){
  scolors=brewer.pal(length(sces), "Set1")
}else{
  scolors=rainbow(length(sces))
}
names(scolors)<-names(sces)

hasSimilarity<-length(sces) > 1
hasDifference<-length(sces) > 2

heatmapKeysize<-0.6
heatmapColors<-bluered(75)
heatmapMargins<-c(10,10)

```

# Overview

scRNABatchQC provides both metrics and diagnostic graphics to assess the similarity/difference on technical factors, expression profiles, and biological features across multiple scRNA-seq experiments, which helps identify potential batch effects. 
scRNABatchQC report include five sections: 

  * _**QC Summary**_ - Overview of Quality Control in each scRNA-seq experiment. 
  * _**Technical View**_ - Technical factors across multiple scRNA-seq experiments. 
  * _**Expression Similarity**_ - Expression Similarity between cells across multiple scRNA-seq experiments. 
  * _**Biological View**_ - Biological features across multiple scRNA-seq experiments.
  * _**Pairwise Difference**_ - Genes and pathways that cause the pairwise difference among scRNA-seq experiments.

# QC Summary

```{r tableSummary, echo=FALSE, results="asis"}
pw<-.getMetaData(sces, "filters")
pw<-cbind(data.frame("EID"=names(sces)), pw)
pw<-subset(pw, select=-c(Sample))
kable(pw, caption=tabRef("tableSummary", "Summary of Quality Metrics."), row.names=FALSE)
```

`r tabRef("tableSummary")` contains a summary of quality metrics generated by the scRNABatchQC package. 

A short description of Table 1 metrics is provided below: 

  * _**EID**_ - Unique experiment ID. 
  * _**Count**_ - The total number of counts 
  * _**Cell**_ - The total number of cells 
  * _**Gene**_ - The total number of genes 
  * _**R-Count**_ - The range of the number of counts across cells [min-median-max]. 
  * _**R-Gene**_ - The range of the number of genes identified in cells [min-median-max] 
  * _**mtRNA**_ - The maximum percentage of mitochondrial RNA reads 
  * _**rRNA**_ - The maximum percentage of rRNA reads
  * _**F-Count**_ - The number of cells filtered by counts 
  * _**F-Gene**_ - The number of cells filtered by genes
  * _**F-rRNA**_ - The number of cells filtered by rRNA percentage
  * _**F-mtRNA**_ - The number of cells filtered by mtRNA percentage 
  * _**F**_ - The total number of cells filtered by low quality

# Technical View

This section checks 11 technical features and aggregates results across multiple scRNA-seq experiments, which include the total number of counts, the number of expressed genes, percentage of reads mapped to mitochondrial RNA and ribosomal RNA, the expression cumulative plot, the detection rate, the mean-variance trend, the variance explained by various technical features.

<hr>

```{r density_total_counts, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("density_total_counts", "Total count distribution"), fig.height=4}
print(plotDensity(sces,"total_counts", "The total number of counts", scolors))
```

`r figRef("density_total_counts")` shows the distribution of the total number of counts in each cell in each scRNA-seq experiment. Different experiments are denoted by different colors. Cells with very low number of read counts, suggesting RNAs are not efficiently captured, are considered as low quality and removed from the downstream analysis. Different count distributions across experiments are indicative of the presence of batch effects.

<hr>

```{r density_total_features, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("density_total_features", "Total feature distribution"), fig.height=4}
print(plotDensity(sces,"total_features", "The number of expressed genes", scolors))
```

`r figRef("density_total_features")` shows the distribution of the number of expressed genes in each cell in each scRNA-seq experiment. Different experiments are denoted by different colors. Cells with very few expressed genes are considered as low quality and removed from the downstream analysis. Different distributions on gene numbers across experiments are indicative of the presence of batch effects.

<hr>

```{r density_pct_counts_Mt, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("density_pct_counts_Mt", "Mitochondrial count distribution"), fig.height=4}
print(plotDensity(sces,"pct_counts_Mt", "The proportion of reads mapped to mitochondrial RNA", scolors))
```

`r figRef("density_pct_counts_Mt")` shows the distribution of the proportion of reads mapped to mitochondrial RNA in each cell in each scRNA-seq experiment. Different experiments are denoted by different colors. A cell with high proportion of reads mapped to mitochondrial RNA is considered as low quality and removed from the downstream analysis. Caution should be paid when one experiment has higher proportion of mitochondrial reads than others.

<hr>

```{r plotRibosomalRNA, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotRibosomalRNA", "The proportion of reads mapped to ribosomal RNA"), fig.height=4}
plot.new()

```

`r figRef("plotRibosomalRNA")` shows the distribution of the proportion of reads mapped to ribosomal RNA in each cell in each scRNA-seq experiment, which is denoted by different color. A cell with high proportion of reads mapped to rRNA is considered as low quality and removed from the downstream analysis. Caution should be paid when one experiment has higher proportion of rRNA reads than others.

<hr>

```{r genesCountdistribution, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("genesCountdistribution", "Cumulative expression plot for the top 500 highly expressed genes"), fig.height=4}
print(plotGeneCountDistribution(sces, scolors))
```

`r figRef("genesCountdistribution")` shows the cumulative expression plot shows the percentage of counts for the top 500 highly expressed genes. The plot is generated based on the average count of each gene, that is, the top 500 highly expressed genes are those with the highest average count. This plot describes how reads are distributed across genes in each scRNA-seq experiment. A cumulative graph with rapid increase at the beginning suggests a library with low complexity where very few genes consume most reads. In contrast, a cumulative graph with slow increase at the beginning indicates that reads are more evenly distributed across genes. Different cumulative expression plots across multiple scRNA-seq experiments are indicative of the presence of batch effects. 

<hr>

```{r aveCountVSdetectRate, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("aveCountVSdetectRate", "The detection rate of genes"), fig.height=4}
print(plotAveCountVSdetectRate(sces, scolors))
```

`r figRef("aveCountVSdetectRate")` shows the percentage of cells expressing each gene plotted against its log-average count. Generally the detection rate for a gene depends on the expression. Different detection rates across scRNA-seq experiments are indicative of the presence of batch effects. 

<hr>

```{r varianceTrend, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("VarianceTrend", "The mean-variance trend")}
print(plotVarianceTrend(sces, scolors), fig.height=4)
```

The variance of gene expression is driven by both uninteresting technical noise and real biological difference. scRNABatchQC models the technical variance of gene expression values using trendVar function, which fits a trend between mean expression abundance and technical variance. The biological variance is then calculated by subtracting the technical component from the total variance of each gene. Different mean-variance trends across scRNA-seq experiments are indicative of the presence of batch effects.

<hr>

```{r log10_total_counts, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("log10_total_counts", "The percentage of expression variances explained by the library size"), fig.height=4}
print(plotMultiSamplesOneExplanatoryVariables(sces, scolors, "log10_total_counts"))
```

`r figRef("log10_total_counts")` shows the percentage of expression variance that is explained by the log-transformed total number of counts for each gene. If percentages are generally small (1-3%), it indicates that gene expressions are not strongly associated with this technical factor. Otherwise, it suggests that the library sizes contribute substantially to the gene expression abundance. If so, this technical factor may need to be modeled into the downstream analysis.  Different contributions of library sizes to expression variances across multiple experiments are indicative of the presence of batch effects.

<hr>

```{r log10_total_features, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("log10_total_features", "The percentage of expression variances explained by the total number of expressed genes"), fig.height=4}
print(plotMultiSamplesOneExplanatoryVariables(sces, scolors, "log10_total_features"))
```

`r figRef("log10_total_features")` shows the percentage of expression variance of each gene that is explained by the log-transformed number of genes detected in each cell. If the percentages are generally small (1-3%), it indicates that the gene expressions are not strongly associated with this technical factor. Otherwise, it suggests that the factor contribute substantially to the gene expression abundance. If so, the technical factor may need to be modeled into the downstream analysis.  Different contributions of the factor to expression variances are indicative of the presence of batch effects.

<hr>

```{r log10_total_counts_Mt, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("log10_total_counts_Mt", "The percentage of expression variances explained by the proportion of mitochondrial reads"), fig.height=4}
print(plotMultiSamplesOneExplanatoryVariables(sces, scolors, "log10_total_counts_Mt"))
```

`r figRef("log10_total_counts_Mt")` shows the percentage of expression variance of each gene that is explained by the percentage of mitochondrial reads. If the percentages are generally small (1-3%), it indicates that the gene expressions are not strongly associated with this technical factor. Otherwise, it suggests that the factor contribute substantially to the gene expression abundance. If so, the technical factor may need to be modeled into the downstream analysis.  Different contributions of the factor to expression variances are indicative of the presence of batch effects.

<hr>

```{r plotVarianceRRNA, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotVarianceRRNA", "The percentage of expression variances explained by the proportion of rRNA reads"), fig.height=4}
plot.new()
```

`r figRef("plotVarianceRRNA")` shows the percentage of expression variance of each gene that is explained by the percentage of mitochondrial reads. If the percentages are generally small (1-3%), it indicates that the gene expressions are not strongly associated with this technical factor. Otherwise, it suggests that the factor contribute substantially to the gene expression abundance. If so, the technical factor may need to be modeled into the downstream analysis.  Different contributions of the factor to expression variances are indicative of the presence of batch effects.

<hr>

```{r child="scRNABatchQCreport_similarity.Rmd", eval=hasSimilarity, echo=FALSE}
```

```{r plotPCA, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotPCA", "PCA plot of expression similarity between cells"), fig.height=4}
print(plotAllPCA(scesall, scolors))
```

The PCA Plot for the first two components (`r figRef("plotPCA")`) shows the relationships between cells within or across scRNA-seq experiments.

<hr>

```{r plotTSNE, eval=DISPLAY_FIGURE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotTSNE", "tSNE plot of expression similarity between cells"), fig.height=4}
print(plotAllTSNE(scesall, scolors))
```

The tSNE plot for the first two components (`r figRef("plotTSNE")`) shows the relationships between cells within or across scRNA-seq experiments. tSNE is a non-linear dimension reduction technique, which PCA reduce the dimension through linear transformation. 

<hr>

```{r child="scRNABatchQCreport_biological.Rmd", eval=hasSimilarity, echo=FALSE}
```

```{r child="scRNABatchQCreport_difference.Rmd", eval=hasDifference, echo=FALSE}
```
